streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -92,69 +95,80 @@ public void validatePartitionExists(final TopicPartition topicPartition, final S
 
     @Override
     public void register(final StateRestorer restorer) {
-        if (restorer.offsetLimit() > 0) {
-            stateRestorers.put(restorer.partition(), restorer);
+        stateRestorers.put(restorer.partition(), restorer);
+    }
+
+
+    public Collection<TopicPartition> restore() {
+        if (!initialized) {
+            initialize();
+        }
+
+        if (needsRestoring.isEmpty()) {
```

Comment:
Thinking about this: I think we do not need this pre-check as in the last call of `restore()` that we return all the partitions, the thread will transit to `RUNNING` already so that whenever `restore()` is called, we know that there is definitely some partitions that still need to restore from.